### PR TITLE
Braincert integration

### PR DESCRIPTION
**Pull request by xyc0562**
BrainCert Whiteboard API integration with Coursemology.

### CODE REVIEW

**Comment by weiqingtoh** [comment]
1. Could you split this up into small commits for review?
2. Can you add feature tests for this?
3. I suspect the API key for integration should be implemented as a setting for each course, rather than a global API key.

**Comment by xyc0562** [comment]
@weiqingtoh 
1. Could you split this up into small commits for review?
Sure. In this case, would it be easier to just close this request and open a new one?

3. I suspect the API key for integration should be implemented as a setting for each course, rather than a global API key.
You are right about putting it into setting, though logically it should be a setting for each instance.

2. Can you add feature tests for this?
An API key is needed for feature tests. Any suggestion how that may be done w/o compromising the key?

File: app/assets/javascripts/course/lectures.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+
```
Comment (allenwq) [inline app/assets/javascripts/course/lectures.js]: @xyc0562 We are moving js files to `client/bundles/`, you should move this file there and use ES6 syntax.

**Comment by weiqingtoh** [comment]
> Sure. In this case, would it be easier to just close this request and open a new one?
No need. What you can do is to update your branch in your fork, and this PR would automatically be updated.

> I suspect the API key for integration should be implemented as a setting for each course, rather than a global API key.
> You are right about putting it into setting, though logically it should be a setting for each instance.

It depends, but the same instance might have courses with different instructors. Each might have a different API key depending on whether they paid and their subscription plan as well. 

> Can you add feature tests for this?
> An API key is needed for feature tests. Any suggestion how that may be done w/o compromising the key?

I think you can explore stubbing the braincert components, so that you can just test the code you have written (and assume that the braincert features work properly).

**Comment by fonglh** [comment]
Agree that the braincert tests should be stubbed, or our tests will take even longer to run because of the network calls.

**Comment by weiqingtoh** [comment]
@xyc0562 the api key need not be an extra fb column, you can implement it as a settings key under the lecture component.

File: app/controllers/system/admin/lectures_controller.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+class System::Admin::LecturesController < System::Admin::Controller
+  load_and_authorize_resource :lecture, class: System::Lecture.name
```
Comment (allenwq) [inline app/controllers/system/admin/lectures_controller.rb]: So we also need a `System::Lecture` ?  system announcement is for announcing important stuff to entire coursemology.

File: app/models/course/lecture.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      return nil, msg
+    end
+    create_classroom unless classroom_id
+    generate_classroom_link user, is_instructor
+  end
+
+  private
+
+  def generate_classroom_link(user, is_instructor)
+    res = call_braincert_api '/v2/getclasslaunch',
```
Comment (allenwq) [inline app/models/course/lecture.rb]: Use `call_braincert_api()` with brackets, our convention only uses DSL form if we don't care about the return value.

File: app/models/course/lecture.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                             generate_classroom_link_params(user, is_instructor)
+    h = JSON.parse res.body
```
Comment (allenwq) [inline app/models/course/lecture.rb]: `JSON.parse(res.body)`

Comment (allenwq) [inline app/models/course/lecture.rb]: Also it's better to have a more meaningful name other than `h`

File: app/models/course/lecture.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      is_recurring: 0,
+      seat_attendees: 25, record: 1, isRegion: 7
+    }
+  end
+
+  def remove_classroom
+    return true unless classroom_id
+    res = call_braincert_api '/v2/removeclass', cid: classroom_id
+    if (error = JSON.parse(res.body)['error'])
+      Rails.logger.error "Error removing classroom for lesson #{id}.\nError is: \n#{error}"
```
Comment (allenwq) [inline app/models/course/lecture.rb]: Seems no one uses  this method, and I think we should feedback to the user instead of logging to the system?

File: app/views/course/admin/admin/index.html.slim [hunk 6]
```diff
@@ -15,6 +15,7 @@
     = f.input :start_at
     = f.input :end_at
     = f.input :gamified, hint: t('.gamified_label')
+    = f.input :braincert_api_key, placeholder: t('.braincert_api_key_hint')
```
Comment (allenwq) [inline app/views/course/admin/admin/index.html.slim]: This field should only be disabled if `Lecture` is enabled, and I think we should move this to the Lecture component settings.

File: app/views/course/lectures/_lecture.html.slim [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    ' by
+    = link_to_user(lecture.creator)
+
+  div.content
+    = format_html(lecture.content)
+
+  div.access-link
+    - link_exists = lecture.instructor_classroom_link.present?
+    - is_instructor = can?(:edit, lecture)
+    - if lecture.currently_active?
```
Comment (allenwq) [inline app/views/course/lectures/_lecture.html.slim]: We have a self-directed learning mode which allows the student access future assignment, should that affect the starting time here ? 

some reference here: https://github.com/Coursemology/coursemology2/blob/master/app/models/course.rb#L121
and you can find the settings under course settings tab.

File: client/app/bundles/course/lectures.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
+$(document).ready(() => {
+  /* global Routes */
```
Comment (allenwq) [inline client/app/bundles/course/lectures.js]: Think we should remove this comment? as in, it's not related.

File: client/app/bundles/course/lectures.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  $(document).on('click', '.gen-access-link', function() {
+    const $this = $(this);
+    const splits = $this.attr('id').split('-');
+    const courseId = splits[1];
+    const lectureId = splits[2];
+    $this.addClass('disabled');
+    $.get(Routes.access_link_course_lecture_path(courseId, lectureId, { format: 'json' }),
```
Comment (allenwq) [inline client/app/bundles/course/lectures.js]: 😄 Thanks for moving this piece of code to the client folder.  But we have certain conventions here also:

All of the client side APIs should be defined and documented in https://github.com/Coursemology/coursemology2/tree/master/client/app/api

( this is quite new, so that we have some old code which stills directly calls the server API and they will be cleaned up finally. so let's follow the convention and move new code there )

File: db/migrate/20170103104020_create_course_lectures.rb [hunk 10]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+class CreateCourseLectures < ActiveRecord::Migration
+  def change
+    create_table :course_lectures do |t|
```
Comment (allenwq) [inline db/migrate/20170103104020_create_course_lectures.rb]: Should the lectures be disabled in Lesson plan page ?  If we need that, the table should use lesson plan as the parent table, otherwise this is fine.

File: app/models/course/lecture.rb [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+  acts_as_readable on: :updated_at
+  has_many_attachments on: :content
+
+  belongs_to :course, inverse_of: :lectures
+
+  scope :sorted_by_date, -> { order(start_at: :desc) }
+  scope :sorted_by_sticky, -> { order(sticky: :desc) }
+
+  def to_partial_path
+    'course/announcements/announcement'
```
Comment (allenwq) [inline app/models/course/lecture.rb]: For allowing future customization, it's better to use it's own partial instead of the announcement one. Otherwise later some guy changes announcement partial and it will surprise him that the lecture also get changed.  what do you think ?

File: app/models/course/lecture.rb [hunk 12]
```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+class Course::Lecture < ActiveRecord::Base
+  acts_as_readable on: :updated_at
+  has_many_attachments on: :content
+
+  belongs_to :course, inverse_of: :lectures
+
+  scope :sorted_by_date, -> { order(start_at: :desc) }
+  scope :sorted_by_sticky, -> { order(sticky: :desc) }
```
Comment (allenwq) [inline app/models/course/lecture.rb]: From the schema definition, there's no sticky attribute. I also think lectures don't need sticky.

File: app/models/course/lecture.rb [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+    action = '/' + action unless action[0] == '/'
+    post "#{BRAINCERT_API_BASE_URL}#{action}",
+         params.merge(apikey: course.braincert_api_key)
+  end
+
+  def create_classroom
+    res = call_braincert_api '/v2/schedule', create_classroom_params
+    h = JSON.parse(res.body)
+    if (error = h['error'])
+      msg = "Error scheduling for lesson #{id}.\nError is: \n#{error}"
```
Comment (allenwq) [inline app/models/course/lecture.rb]: Should this error be populated to the user ?

File: app/views/course/lectures/index.html.slim [hunk 14]
```diff
@@ -0,0 +1,6 @@
+= page_header lectures_title do
+  = new_button([current_course, :lecture]) if can?(:create, Course::Lecture.new(course: current_course))
+
+= render partial: 'course/lectures/lecture', collection: @lectures,
```
Comment (allenwq) [inline app/views/course/lectures/index.html.slim]: Remove the `to_parital_path` in Lecture model, and you only need `render @lectures`

File: app/views/course/lectures/index.html.slim [hunk 15]
```diff
@@ -0,0 +1,6 @@
...
+  spacer_template: 'announcements/announcement_spacer'
```
Comment (allenwq) [inline app/views/course/lectures/index.html.slim]: The lecture should use its own spacer too

**Comment by allenwq** [comment]
Agree with @weiqingtoh that the key should be defined in Component settings

File: app/controllers/system/admin/lectures_controller.rb (hunk 2 above)
Comment (xyc0562) [inline app/controllers/system/admin/lectures_controller.rb]: Will remove

File: app/models/course/lecture.rb (hunk 12 above)
Comment (xyc0562) [inline app/models/course/lecture.rb]: Will remove

File: app/models/course/lecture.rb (hunk 11 above)
Comment (xyc0562) [inline app/models/course/lecture.rb]: Sure

File: app/models/course/lecture.rb (hunk 3 above)
Comment (xyc0562) [inline app/models/course/lecture.rb]: Sure

File: app/models/course/lecture.rb (hunk 4 above)
Comment (xyc0562) [inline app/models/course/lecture.rb]: OK

File: app/models/course/lecture.rb (hunk 13 above)
Comment (xyc0562) [inline app/models/course/lecture.rb]: It depends. But for the sake of consistency, I will propagate this to the front-end

File: app/models/course/lecture.rb (hunk 5 above)
Comment (xyc0562) [inline app/models/course/lecture.rb]: Will remove this method altogether

File: app/views/course/lectures/_lecture.html.slim (hunk 7 above)
Comment (xyc0562) [inline app/views/course/lectures/_lecture.html.slim]: It doesn't matter

File: client/app/bundles/course/lectures.js (hunk 8 above)
Comment (xyc0562) [inline client/app/bundles/course/lectures.js]: Sure

File: client/app/bundles/course/lectures.js (hunk 9 above)
Comment (xyc0562) [inline client/app/bundles/course/lectures.js]: OK

File: db/migrate/20170103104020_create_course_lectures.rb (hunk 10 above)
Comment (xyc0562) [inline db/migrate/20170103104020_create_course_lectures.rb]: There is no need to for lectures to be in Lesson Plan page

File: app/views/course/lectures/index.html.slim (hunk 14 above)
Comment (xyc0562) [inline app/views/course/lectures/index.html.slim]: OK

File: app/views/course/lectures/index.html.slim (hunk 15 above)
Comment (xyc0562) [inline app/views/course/lectures/index.html.slim]: OK

**Comment by weiqingtoh** [comment]
@xyc0562 Actually given the size, I would suggest maybe that you can do PRs in smaller batches, ie. models, controllers+views. Let's fix the models first so you don't have to keep changing controllers + views every time we suggest changes. 

Try to keep PRs to around 300-400 lines and we can review and you can change quickly.

File: app/models/course.rb [hunk 16]
```diff
@@ -126,6 +127,14 @@ def advance_start_at_duration=(time)
     settings(:course).advance_start_at_duration = time
   end
 
+  def braincert_whiteboard_api_key
```
Comment (allenwq) [inline app/models/course.rb]: Since you already defined `app/models/course/lecture_settings.rb`, they key should be set there as well?

Comment (xyc0562) [inline app/models/course.rb]: Done

**Comment by allenwq** [comment]
@xyc0562 the building is failing. Need to fix the hound conflicts though.

😄  Also if you don't mind can add newlines to the end of the files ? 

@weiqingtoh do you have more comments ?

File: app/controllers/components/course/lectures_component.rb [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+
+  private
+
+  def main_sidebar_items
+    [
+      {
+        key: :lectures,
+        icon: 'television',
+        title: settings.title || t('course.lectures.sidebar_title'),
+        weight: 1,
```
Comment (weiqingtoh) [inline app/controllers/components/course/lectures_component.rb]: You might want to compare this weight with the rest - do you want lectures to be at the top? By default we usually leave Announcements at the top.

File: app/controllers/course/lectures_controller.rb [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+          render json: { errors: errors }, status: 400
+        else
+          render json: { link: link }
+        end
+      end
+    end
+  end
+
+  def index #:nodoc:
+    @lectures = @lectures.includes(:creator).sorted_by_date
```
Comment (weiqingtoh) [inline app/controllers/course/lectures_controller.rb]: Can combine into `@lectures = @lectures.includes(:creator).sorted_by_date.page(page_param)`

File: app/models/components/course/lectures_ability_component.rb [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+      allow_staff_manage_lectures
+    end
+
+    super
+  end
+
+  private
+
+  def allow_students_show_lectures
+    can :read, Course::Lecture,
```
Comment (weiqingtoh) [inline app/models/components/course/lectures_ability_component.rb]: `can [:read, :access_link], Course::Lecture, course_all_course_users_hash`

File: app/models/course/lecture.rb [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+
+  belongs_to :course, inverse_of: :lectures
+
+  scope :sorted_by_date, -> { order(start_at: :desc) }
+
+  TIME_BRAINCERT = '%I:%M%p'
+  DATE_ISO = '%Y-%m-%d'
+  BRAINCERT_API_BASE_URL = 'https://api.braincert.com'
+
+  def handle_access_link(user, is_instructor)
```
Comment (weiqingtoh) [inline app/models/course/lecture.rb]: I would think that `handle_access_link` should be placed in a service then called in the controller, together with:
 - `generate_classroom_link`
 - `generate_classroom_link_params`
 - `post`
 - `call_braincert_api`
 - `create_classroom`
 - `create_classroom_params`

File: app/models/course/lecture.rb [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+    end
+  end
+
+  def post(url, params)
+    Net::HTTP.post_form URI.parse(url), params
+  end
+
+  def call_braincert_api(action, params)
+    action = '/' + action unless action[0] == '/'
+    settings = Course::LectureSettings.new(course.settings(:lecture))
```
Comment (weiqingtoh) [inline app/models/course/lecture.rb]: If you move this into the service and call it in your controller, this key would be preloaded already.

File: app/models/course/lecture_settings.rb [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+  # @param [String] title The new title
+  def title=(title)
+    title = nil unless title.present?
+    @settings.title = title
+  end
+
+  # Returns the lecture pagination count
+  #
+  # @return [Integer] The pagination count of lecture
+  def pagination
```
Comment (weiqingtoh) [inline app/models/course/lecture_settings.rb]: How many lectures do you expect per course - do you really need pagination?

File: app/models/course/lecture.rb [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+    else
+      I18n.t(:'course.lectures.lesson_already_conducted')
+    end
+  end
+
+  def post(url, params)
+    Net::HTTP.post_form URI.parse(url), params
+  end
+
+  def call_braincert_api(action, params)
```
Comment (weiqingtoh) [inline app/models/course/lecture.rb]: Would you prefer to use javascript to do this parts? Feels like javascript-ish code in ruby-style. But if you want to leave it like that I think it's fine, but harder to maintain and harder to handle callbacks.

File: config/routes.rb [hunk 24]
```diff
@@ -146,6 +149,10 @@
         end
       end
 
+      resources :lectures, concerns: :paginatable do
+        get :access_link, on: :member
```
Comment (weiqingtoh) [inline config/routes.rb]: Don't need the `on: :member` if I remember correctly. You can remove it and verify by running `rake routes`.

File: app/helpers/course/lectures_helper.rb [hunk 25]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+module Course::LecturesHelper
+  # Returns the formatted title of lectures component.
+  #
+  # @return [String] The formatted title of lectures component.
+  # @return [nil] If the title is nil.
+  def lectures_title
+    @settings.title.nil? ? nil : format_inline_text(@settings.title)
```
Comment (weiqingtoh) [inline app/helpers/course/lectures_helper.rb]: Don't you want to return the default title if there is no title specified in the settings?

File: app/controllers/components/course/lectures_component.rb (hunk 17 above)
Comment (xyc0562) [inline app/controllers/components/course/lectures_component.rb]: It looks like all consecutive integers are used. Let's say I want to put lecture directly below announcement. Should I:

1. Use decimal like 1.5?
2. Change all other components' weights in component.rb?

File: config/routes.rb (hunk 24 above)
Comment (xyc0562) [inline config/routes.rb]: It doesn't work. W/O `on: :member` `load_and_authorize_resource` doesn't work correctly in lectures_controller (`@lecture` doesn't gets loaded)

**Comment by xyc0562** [comment]
@allenwq @weiqingtoh Any other comments?

File: client/app/api/course/Assessments.js [hunk 26]
```diff
@@ -11,7 +11,7 @@ export default class AssessmentsAPI extends BaseCourseAPI {
   * error response: { errors: [] } - An array of errors will be returned upon validation error.
   */
   update(assessmentId, params) {
-    return this.getClient().patch(`${this._getUrlPrefix()}/${assessmentId}`, params);
+    return this.getClient().patch(`${this._geturlprefix()}/${assessmentid}`, params);
```
Comment (weiqingtoh) [inline client/app/api/course/Assessments.js]: Why is this changed?

File: config/locales/en/course/admin/lecture_settings.yml [hunk 27]
```diff
@@ -0,0 +1,11 @@
+en:
+  course:
+    admin:
+      lecture_settings:
+        edit:
+          title_placeholder: 'Leave blank to use default title'
+          braincert_whiteboard_api_key_hint: >
+            For Lecture component
```
Comment (weiqingtoh) [inline config/locales/en/course/admin/lecture_settings.yml]: `Braincert Whiteboard API Key for the Lecture Component`

File: client/app/api/course/Assessments.js (hunk 26 above)
Comment (xyc0562) [inline client/app/api/course/Assessments.js]: Ooops.

It's a mistake. Sorry about that

File: config/locales/en/course/lectures.yml [hunk 28]
```diff
@@ -0,0 +1,27 @@
...
+      create:
+        success: 'Lecture %{title} was created.'
+      edit:
+        header: 'Edit Lecture'
+      update:
+        success: 'Lecture %{title} was updated.'
+      destroy:
+        success: 'Lecture %{title} was deleted.'
+        failure: 'Failed to delete lecture: %{error}'
+      lesson_already_conducted: "lesson has already been conducted"
```
Comment (weiqingtoh) [inline config/locales/en/course/lectures.yml]: `The lesson has already ended`

File: config/locales/en/course/lectures.yml [hunk 29]
```diff
@@ -0,0 +1,27 @@
...
+      lesson_live_in: "lesson will go live in %{desc}, please refresh the page when it is live"
```
Comment (weiqingtoh) [inline config/locales/en/course/lectures.yml]: `The lesson will go live in %{desc}, please refresh the page when it is live`

File: config/routes.rb (hunk 24 above)
Comment (weiqingtoh) [inline config/routes.rb]: 👍

**Comment by weiqingtoh** [comment]
Do fix the newlines at the end of the file as well. Also, it would be good to write some tests after this to ensure that refactoring doesn't break anything.

Otherwise, overall some parts can be better structured (within the files) but can be merged given that it is urgent.

**Comment by allenwq** [comment]
@xyc0562  The lint still fails, you can run `cd client && yarn lint` to find the exact errors. there's a ` yarn lint-fix` too which can help you auto fix some of the errors.

**Comment by xyc0562** [comment]
@weiqingtoh @allenwq 

Thanks for the time and effort in reviewing my (not so professional) code. Hopefully my future pull requests (if there is to be any) will be easier to for you guys.

Cheers and good night :-)

**Comment by allenwq** [comment]
Thanks @xyc0562 ! good night 😄

**Comment by weiqingtoh** [comment]
No problems - thanks for being patient with our 'OCD-ness' @xyc0562
